trailmap/pathway_analysis.py

```diff
@@ -0,0 +1,319 @@
+        "MultiDiGraphs, the graph form that Trailmap uses. Graph will be " +
+        "converted to DiGraph, which can lose information if you have " +
+        "multiple edges running between two nodes. \nChecking if graph G " +
+        "has an identical DiGraph representation...\n")
+
+    if G.is_directed():
+        multiedges = has_multiedges(G)
+        if multiedges == True:
+            print(warning_message)
+            print("G appears to have multiple edges between at least one " +
+            "pair of nodes. Information WILL be lost in the switch to " +
+            "DiGraph. Proceeding with caution, consider the results at " +
+            "your own risk\n")
+            H = nx.DiGraph(G)
+        elif multiedges == False:
+            print(warning_message)
+            print("G does not appear to have multiple edges between any " +
+            "pair of nodes. No information should be lost in the switch to " +
+            "DiGraph. Proceeding with flow calculation.\n")
+            H = nx.DiGraph(G)
+        else:
+            H=G
+    else:
+        print("Provided graph is not a MultiDiGraph or DiGraph and other \
+        errors may occur. Abort.\n")
+        return None, None, G
+
+    max_flow_path = nx.maximum_flow(H, s, t)
+    max_flow = nx.maximum_flow_value(H, s, t)
+    return max_flow_path, max_flow, H
+
+
+def find_pathway_flow(G, pathway):
+    '''returns the maximum permissible flow for a given pathway. Any edge
+    without 'capacity' attribute will be given infinite capacity.
+    '''
+    multiedges = has_multiedges(G)
+    if multiedges != True:
+        # graph either has no multiedges, or is not in MultiDiGraph format
+        H = nx.DiGraph(G)
+        edges = set(pairwise(pathway))
+        #H.remove_nodes_from([n for n in G if n not in pathway])
+        H_sg = H.edge_subgraph(edges).copy()
+
+        sources = get_sources(H_sg)
+        sinks = get_sinks(H_sg)
+
+        # check for capacity in edges (if no capacity, NetworkX will raise an
+        # exception) and if every edge has infinite flow, return inf
+        infinite_flow = True
+        for edge in edges:
+            if 'capacity' not in H_sg.get_edge_data(*edge):
+                return math.inf
+            if math.isinf(H_sg.get_edge_data(*edge)['capacity']) is False:
+                infinite_flow = False
+            
+        if infinite_flow is True:
+            return math.inf
+        else:
+            path_flow = nx.maximum_flow_value(H_sg, sources[0], sinks[0])
+            return path_flow
+
+    else:
+        print("Provided pathway has multiple edges. Flow calculation across \
+               multi-edged pathways is currently unsupported. Returning None")
+        return None
+
+
+def find_simple_cycles(G): # pragma: no cover
+    '''finds cycles in a graph and returns them in a list of lists.
+    '''
+    sc = list(nx.simple_cycles(G))
+    return sc
+
+
+def check_if_sublist(path, set_of_steps):
```

Comment:
What is the form of `set_of_steps`?

Reply:
should be a list, so calling it a set is a silly idea...